lyricsheets/models/karaoke.py

```diff
@@ -118,6 +233,51 @@ def calculate_char_fade_offsets(self, style: Style, transitionDuration: timedelt
             char.fadeOffset = charAccTime
             charAccTime += lineCharTime
 
+    def calculate_sizing(self, style: Style):
```

Comment:
The design of this method kinda breaks the class encapsulation, in that this method calculates the size of a line assuming it is being rendered in a particular style, but the `KLine` object itself doesn't contain a style.

So if the object is passed around, how would one be able to determine what style all the calculations correspond to? (Less importantly, but) what about if the text is changed?

Suggestions:
1) Move the style to the constructor so every KLine object represents a particular line in a particular style, and either do the sizing calculations in the constructor itself or lazily it is needed later.
2) Not terribly important, but maybe consider using https://docs.python.org/3/library/dataclasses.html#frozen-instances to make these classes immutable

In general, not a fan of designs where users need to "remember to call a certain method before using these values or else you may get invalid results" since one day, someone will forget. Might as well ensure that the values of everything are always correct by encapsulating that logic within the class itself.

If you prefer to keep the Style out of these classes and allow users to calculate the sizes of the same object under different styles, a better way would to have this method actually return a `Sizing` object that contains all the relevant sizing fields, which would be fine as well.

Reply:
> So if the object is passed around, how would one be able to determine what style all the calculations correspond to?

Right now, the style is specified by the script that generates the events for the effects, so the script will be the one keeping track of it. The onus is on the script writer to keep track of what style they want to use for a particular line and not assume the wrong style.

> What about if the text is changed?

I don't think this will be an issue specific to styles. Changing the text should be done before generating any positional information. Once the positional information is calculated, the text should be finalized.

The second suggestion would make it unwieldy needing to keep track of the sizing for the lines, syllables and characters, and store them in such a way that you can refer to the syllable/character's sizing and existing fields unless we duplicate all the existing information when creating the `Sizing` fields. I think that's fine, but creating new classes for that reason sounds a little bit wasteful and unnecessary.

> In general, not a fan of designs where users need to "remember to call a certain method before using these values or else you may get invalid results" since one day, someone will forget.

How about this? Update `KaraokeEffect` (or maybe `Effect` for `NoKaraokeEffect`?) to add in a `get_style()` abstract method, which takes in a `SongLine` or `KLine` and returns a `Style` that can be chosen based on whatever the inherited class desires. Or it could even just default to  ROMAJI_STYLE and EN_STYLE, as is similar to what's currently in `DefaultEffect`. Then, have the `KaraokeEffect`'s `to_romaji_events()` and `to_en_events()` call the constructor of `KLine` with the `Style` that is returned, generating all the necessary information for the `KLine`, `KSyl` and `KChar` before it is passed to `to_romaji_k_events()` and `to_en_k_events()`, where it is then ready to go.  This keeps the separation between `SongLine` and `KLine` (the inherited classes don't have to know what a `SongLine` is) and ensures that the script writer will never forget.

Reply:
Now that you've described the entire process, it does seem a little clunky......

After thinking about it again I think my gripe is more with not being able to clearly tell what style is bound to a `K*` object and forgetting to call `calculate_sizings` might be an easily-missed error

If this change makes things too clunky I think it'd be fine if we could (from the original commit):
- Add a `style` field to all `K*` classes
- Change the method name of `calculate_sizings` to `bind_style`
- Ideally, make all extent-related fields throw an Exception if they're accessed before a style is bound to the `K*` object

Agreed that we can ignore the changing text thing though.

Reply:
Updated with suggestions.